tests/github/interactive_api/single_file_experiments/pytorch_kvasir_unet.py

```diff
@@ -0,0 +1,387 @@
+import logging
+
+import torch
+import torch.nn as nn
+import torch.optim as optim
+
+from tests.github.interactive_api.layers import soft_dice_loss, soft_dice_coef, double_conv, down, \
+    up
+
+logger = logging.getLogger(__name__)
+
+"""
+UNet model definition
+"""
+
+
+class UNet(nn.Module):
+    def __init__(self, n_channels=3, n_classes=1):
+        super().__init__()
+        self.inc = double_conv(n_channels, 64)
+        self.down1 = down(64, 128)
+        self.down2 = down(128, 256)
+        self.down3 = down(256, 512)
+        self.down4 = down(512, 1024)
+        self.up1 = up(1024, 512)
+        self.up2 = up(512, 256)
+        self.up3 = up(256, 128)
+        self.up4 = up(128, 64)
+        self.outc = nn.Conv2d(64, n_classes, 1)
+
+    def forward(self, x):
+        x1 = self.inc(x)
+        x2 = self.down1(x1)
+        x3 = self.down2(x2)
+        x4 = self.down3(x3)
+        x5 = self.down4(x4)
+        x = self.up1(x5, x4)
+        x = self.up2(x, x3)
+        x = self.up3(x, x2)
+        x = self.up4(x, x1)
+        x = self.outc(x)
+        x = torch.sigmoid(x)
+        return x
+
+
+model_unet = UNet()
+
+optimizer_adam = optim.Adam(model_unet.parameters(), lr=1e-4)
+
+import os
+from hashlib import sha384
+import PIL
+from torch.utils.data import Dataset, DataLoader
+from torchvision import transforms as tsf
+from skimage import io
+
+os.makedirs('data', exist_ok=True)
+os.system(
+    "wget -nc 'https://datasets.simula.no/hyper-kvasir/hyper-kvasir-segmented-images.zip' -O ./data/kvasir.zip")
+ZIP_SHA384 = 'e30d18a772c6520476e55b610a4db457237f151e' \
+             '19182849d54b49ae24699881c1e18e0961f77642be900450ef8b22e7'
+assert sha384(open('./data/kvasir.zip', 'rb').read(
+    os.path.getsize('./data/kvasir.zip'))).hexdigest() == ZIP_SHA384
+os.system('unzip -n ./data/kvasir.zip -d ./data')
+
+DATA_PATH = './data/segmented-images/'
+import numpy as np
+
+
+def read_data(image_path, mask_path):
+    """
+    Read image and mask from disk.
+    """
+    img = io.imread(image_path)
+    assert (img.shape[2] == 3)
+    mask = io.imread(mask_path)
+    return (img, mask[:, :, 0].astype(np.uint8))
+
+
+class KvasirDataset(Dataset):
+    """
+    Kvasir dataset contains 1000 images for all collaborators.
+    Args:
+        data_path: path to dataset on disk
+        collaborator_count: total number of collaborators
+        collaborator_num: number of current collaborator
+        is_validation: validation option
+    """
+
+    def __init__(self, images_path='./data/segmented-images/images/',
+                 masks_path='./data/segmented-images/masks/',
+                 validation_fraction=1 / 8, is_validation=False):
+
+        self.images_path = images_path
+        self.masks_path = masks_path
+        self.images_names = [img_name for img_name in sorted(os.listdir(
+            self.images_path)) if len(img_name) > 3 and img_name[-3:] == 'jpg']
```

Comment:
self.images_names = [
    img_name 
    for img_name in sorted(os.listdir(self.images_path)) 
    if len(img_name) > 3 and img_name[-3:] == 'jpg'
]